### PR TITLE
[WIP][Parser][Qol] improve diagnostic with missing "self." in init

### PR DESCRIPTION
**Pull request by dduan**
#### What's in this pull request?

before:

```
$ cat t.swift
class A {
    init(x: Int) {}

    convenience init() { init(x: 1) }
}

$ swiftc -c t.swift
t.swift:6:13: error: initializers may only be declared within a type
    init(x: 1)
        ^
t.swift:6:17: error: expected parameter type following ':'
    init(x: 1)
            ^
t.swift:6:17: error: expected ',' separator
    init(x: 1)
            ^
           ,
t.swift:6:17: error: expected parameter type following ':'
    init(x: 1)
            ^
t.swift:6:17: error: expected ',' separator
    init(x: 1)
            ^
                                                                                                                                                       ,
```

after:

```
t.swift:6:9: error: missing 'self.' at initializer invocation
    init(x: 1)
    ^
    self.
```
#### Resolved bug number: ([SR-852](https://bugs.swift.org/browse/SR-852))

---

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>

The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

### CODE REVIEW

**Comment by ddunbar** [comment]
@swift-ci please test

**Comment by kballard** [comment]
What about the following case?

```swift
class Foo {
  init(x: Int) {}
}

class Bar: Foo {
  init() {
    init(x: 1)
  }
}
```

This is going to suggest the insertion of `self.`, but since this is a designated initializer instead of a convenience one, presumably we should be suggesting the insertion of `super.` instead.

**Comment by dduan** [comment]
Good point Kevin. I'll look into what's possible here.

**Comment by dduan** [comment]
Ok, updated so that `.super` will be reported and suggested. Tests are passing locally.

Please @swift-ci please test 😸

File: lib/Parse/ParseStmt.cpp [hunk 1]
```diff
@@ -410,22 +410,32 @@ ParserStatus Parser::parseBraceItems(SmallVectorImpl<ASTNode> &Entries,
...
-        // If this is a normal library, you can't have expressions or statements
-        // outside at the top level.
-        diagnose(StartLoc,
-                 Result.is<Stmt*>() ? diag::illegal_top_level_stmt
-                                    : diag::illegal_top_level_expr);
-        Result = ASTNode();
-      }
+      if (Tok.is(tok::kw_init)) {
+        if (auto CD = dyn_cast<ConstructorDecl>(CurDeclContext)) {
+          // hint at missing 'self.' or 'super.' then skip this statement
```
Comment (kballard) [inline lib/Parse/ParseStmt.cpp]: Comment should be a full sentence, including capitalization and punctuation.

**Comment by kballard** [comment]
LGTM beyond the noted comment.

**Comment by tkremenek** [comment]
@swift-ci please test

**Pull request merged**